packages/flutter/lib/src/widgets/drag_target.dart

```diff
@@ -524,9 +524,12 @@ class _DragTargetState<T> extends State<DragTarget<T>> {
         _candidateAvatars.add(avatar);
       });
       return true;
+    } else {
```

Comment:
Should we also then remove the rejected avatar in `didDrop` (l#547)? It looks like `didLeave` (l#535) removes the avatar from rejected and candidates, but only candidates are removed in `didDrop`. 
This might result in an assert being thrown if you drag and drop a rejected avatar once, and then again since the assert above (l#521) checks that it is not already in the list.

Reply:
If that is the case I would add another test that drags and drops rejected targets twice, as well as drags and leaves for good measure, to check that doesn't trigger the assertion.

Reply:
Thanks for the review! 

So didDrop is never called with an avatar in _rejectedAvatars, only ones in _acceptedAvatars, hence this assertion will never fail. This is because didDrop is only called on _DragAvatar._activeTarget (#690) and _activeTarget is guaranteed to be a target which is accepting the avatar as it is only ever a target which returns true on _DragTargetState.didEnter (#661, #666), which only returns true if it is accepted (#526). Phew. 

Given this long chain of reasoning closely tied to maintaining internal state in multiple widgets I agree it's prudent to introduce some extra tests.

I've updated my original test to be the two you suggested:
* Check that DragTarget rebuilds correctly when the same rejected Draggable is dropped two times in a row.
* Check that DragTarget rebuilds correctly when a rejected Draggable leaves then enters.

Reply:
Ah! Thank you for being so thorough! 
I think I mixed up a re-factor I've been looking into to address #37053 that would involve `didDrop`. 😜 
Thank you for adding the additional test as well. This is a great fix - which I see you mentioned (in #38786) may eradicate the need for exposing another callback in #37053. I'd be interested to see feedback there after this lands.